metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java

```diff
@@ -242,16 +242,24 @@ void publishLogDelta(
         bytesSinceLastSnapshot += manifest.numBytes();
         if (bytesSinceLastSnapshot >= maxBytesSinceLastSnapshot) {
             if (eventQueue.isEmpty()) {
-                scheduleEmit("we have replayed at least " + maxBytesSinceLastSnapshot +
-                    " bytes", newImage);
+                if (manifest.provenance().isOffsetBatchAligned()) {
+                    scheduleEmit("we have replayed at least " + maxBytesSinceLastSnapshot +
+                        " bytes", newImage);
+                } else if (log.isDebugEnabled()) {
+                    log.debug("Not scheduling snapshot because last contained offset is not aligned to a batch boundary.");
```

Comment:
I'm also going to move the call to `resetSnapshotCounters` to the `else` block in `maybeScheduleEmit`. Previously the only way we enter this method and not create a snapshot is during shutdown, so the time/bytes since last snapshot doesn't really matter. Now, since we don't publish a snapshot when it's not batch aligned, we shouldn't reset these values, since we may be over the threshold and want to snapshot whenever our offset is batch aligned once more.